system/emergency_goal_manager/src/emergency_goal_manager_core.cpp

```diff
@@ -0,0 +1,150 @@
+// limitations under the License.
+
+#include <emergency_goal_manager_core.hpp>
+
+namespace emergency_goal_manager
+{
+EmergencyGoalManager::EmergencyGoalManager() : Node("emergency_goal_manager")
+{
+  // Subscriber
+  sub_emergency_goals_ = create_subscription<tier4_system_msgs::msg::EmergencyGoalsStamped>(
+    "~/input/emergency_goals", rclcpp::QoS{1},
+    std::bind(&EmergencyGoalManager::onEmergencyGoals, this, std::placeholders::_1));
+  sub_emergency_goals_clear_command_ =
+    create_subscription<tier4_system_msgs::msg::EmergencyGoalsClearCommand>(
+      "~/input/emergency_goals_clear_command", rclcpp::QoS{1},
+      std::bind(&EmergencyGoalManager::onEmergencyGoalsClearCommand, this, std::placeholders::_1));
+
+  // Client
+  client_set_mrm_route_points_callback_group_ =
+    create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  client_set_mrm_route_points_ = create_client<SetRoutePoints>(
+    "/planning/mission_planning/mission_planner/srv/set_mrm_route",
+    rmw_qos_profile_services_default, client_set_mrm_route_points_callback_group_);
+  client_clear_mrm_route_callback_group_ =
+    create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  client_clear_mrm_route_ = create_client<ClearRoute>(
+    "/planning/mission_planning/mission_planner/srv/clear_mrm_route",
+    rmw_qos_profile_services_default, client_clear_mrm_route_callback_group_);
+
+  // Initialize
+  while (!client_set_mrm_route_points_->wait_for_service(std::chrono::seconds(1)) && rclcpp::ok()) {
+  }
+  while (!client_clear_mrm_route_->wait_for_service(std::chrono::seconds(1)) && rclcpp::ok()) {
+  }
+}
+
+void EmergencyGoalManager::onEmergencyGoals(
+  const tier4_system_msgs::msg::EmergencyGoalsStamped::SharedPtr msg)
+{
+  if (!emergency_goals_map_.empty()) {
+    emergency_goals_map_.clear();
+  }
+
+  std::queue<geometry_msgs::msg::Pose> emergency_goals_queue;
+  for (const auto & goal : msg->goals) {
+    emergency_goals_queue.push(goal);
+  }
+  emergency_goals_map_.emplace(msg->sender, emergency_goals_queue);
+
+  callSetMrmRoutePoints();
+}
+
+void EmergencyGoalManager::onEmergencyGoalsClearCommand(
+  const tier4_system_msgs::msg::EmergencyGoalsClearCommand::SharedPtr msg)
+{
+  if (emergency_goals_map_.count(msg->sender) == 0) {
+    RCLCPP_WARN(get_logger(), "Emergency goals from %s is empty.", msg->sender.c_str());
+  }
+
+  emergency_goals_map_.erase(msg->sender);
+
+  if (emergency_goals_map_.empty()) {
+    callClearMrmRoute();
+  } else {
+    callSetMrmRoutePoints();
+  }
+}
+
+void EmergencyGoalManager::callSetMrmRoutePoints()
+{
+  auto request = std::make_shared<SetRoutePoints::Request>();
+  request->header.frame_id = "map";
+  request->header.stamp = this->now();
+  request->option.allow_goal_modification = true;
+
+  while (!emergency_goals_map_.empty()) {
+    // TODO(Makoto Kurihara): set goals with the highest priority
+    auto goals = emergency_goals_map_.begin();
```

Comment:
This node is executed by MultipleExecutor, so emergency_goals_map_ need to get a lock.

Reply:
Sorry, I was mistaken.
This call back function is executed mutually exclusive by default.